proof/crefine/RISCV64/IpcCancel_C.thy

```diff
@@ -2935,6 +3043,88 @@ lemma reply_remove_ccorres:
 
+lemma getBlockingObject_BlockedOnReceive_return:
+  "(getBlockingObject (Structures_H.thread_state.BlockedOnReceive oref cg ro)) = return oref"
+  unfolding getBlockingObject_def epBlocked_def by simp
+
+lemma getBlockingObject_BlockedOnSend_return:
+  "(getBlockingObject (Structures_H.thread_state.BlockedOnSend oref badge cg cgr isc)) = return oref"
+  unfolding getBlockingObject_def epBlocked_def by simp
+
+lemma BlockedOnReceive_replyObject_no_0:
+  "\<lbrakk> ts = BlockedOnReceive oref cg ro; no_0_obj' s; valid_tcb_state' ts s \<rbrakk>
+   \<Longrightarrow> ro \<noteq> Some 0"
+  by (auto simp: valid_tcb_state'_def)
+
+lemma ctcb_relation_BlockedOnReceive_blockingObject:
+  "\<lbrakk> BlockedOnReceive oref cg ro = tcbState tcb; ctcb_relation tcb ctcb \<rbrakk>
+   \<Longrightarrow> blockingObject_CL (thread_state_lift (tcbState_C ctcb)) = oref"
+  by (cases "(tcbState tcb)", simp_all add: ctcb_relation_def cthread_state_relation_def)
+
+lemma ctcb_relation_BlockedOnReceive_replyObject:
+  "\<lbrakk> BlockedOnReceive oref cg ro = tcbState tcb; ctcb_relation tcb ctcb \<rbrakk>
+   \<Longrightarrow> replyObject_CL (thread_state_lift (tcbState_C ctcb)) = option_to_0 ro"
+  by (cases "(tcbState tcb)", simp_all add: ctcb_relation_def cthread_state_relation_def)
+
+lemma ctcb_relation_BlockedOnSend_blockingObject:
+  "\<lbrakk> BlockedOnSend oref badge cg cgr isc = tcbState tcb; ctcb_relation tcb ctcb \<rbrakk>
+   \<Longrightarrow> blockingObject_CL (thread_state_lift (tcbState_C ctcb)) = oref"
+  by (cases "(tcbState tcb)", simp_all add: ctcb_relation_def cthread_state_relation_def)
+
+lemma BlockedOnReceive_ep_at':
+  "\<lbrakk> st_tcb_at' ((=) (BlockedOnReceive oref cg ro)) t s; valid_objs' s \<rbrakk>
+   \<Longrightarrow> ep_at' oref s"
+  by (fastforce dest!: tcb_in_valid_state' simp: obj_at'_def valid_tcb_state'_def)
+
+lemma BlockedOnSend_ep_at':
+  "\<lbrakk> st_tcb_at' ((=) (BlockedOnSend oref badge cg cgr isc)) t s; valid_objs' s \<rbrakk>
+   \<Longrightarrow> ep_at' oref s"
+  by (fastforce dest!: tcb_in_valid_state' simp: obj_at'_def valid_tcb_state'_def)
+
+lemma sym_ref_BlockedOnReceive_replyObject_linked:
+  assumes st: "st_tcb_at' ((=) (Structures_H.thread_state.BlockedOnReceive oref cg (Some ro))) thread s"
+  assumes sy: "sym_refs (state_refs_of' s)"
+  assumes vo: "valid_objs' s"
+  shows "obj_at' (\<lambda>reply. replyTCB reply = Some thread) ro s"
+proof -
+  from vo and st have ra:
+    "reply_at' ro s"
+    by - (drule (1) tcb_in_valid_state', fastforce simp: obj_at'_def valid_tcb_state'_def)
+  from st have ko:
+    "\<exists>tcb. ko_at' tcb thread s \<and> tcbState tcb = BlockedOnReceive oref cg (Some ro)"
+    by (clarsimp simp: st_tcb_at'_def obj_at'_def)
+  with sy and vo and st and ra show ?thesis
+    apply (clarsimp simp: st_tcb_at'_def)
+    apply (frule (1) sym_ref_Receive_or_Reply_replyTCB')
+    apply simp
+    apply (clarsimp simp: obj_at'_def)
+    done
+qed
+
+lemma cancelIPC_threadSet_wp:
```

Comment:
It seems unusual to have this separate from the main proof. Is there any particular reason it is here? I think it could be put into the main proof pretty easily. 

And I thought that `sym_refs_asrt_def` was in the global `[simp]` set, so you shouldn't need to unfold it. But if not, it may be easiest to write `unfolding sym_refs_asrt_def` at the beginning of the main proof, which should unfold it everywhere for you.